frontend/catalyst/from_plxpr/decompose.py

```diff
@@ -0,0 +1,389 @@
+        """Interpret a PennyLane operation instance.
+
+        Args:
+            op (Operator): a pennylane operator instance
+
+        Returns:
+            Any
+
+        This method is only called when the operator's output is a dropped variable,
+        so the output will not affect later equations in the circuit.
+
+        We cache the list of operations seen during the interpretation
+        to build the decomposition graph in the later stages.
+
+        See also: :meth:`~.interpret_operation_eqn`.
+
+        """
+
+        self._operations.add(op)
+        data, struct = jax.tree_util.tree_flatten(op)
+        return jax.tree_util.tree_unflatten(struct, data)
+
+    def interpret_measurement(self, measurement: "qml.measurement.MeasurementProcess"):
+        """Interpret a measurement process instance.
+
+        Args:
+            measurement (MeasurementProcess): a measurement instance.
+
+        See also :meth:`~.interpret_measurement_eqn`.
+
+        """
+
+        # If we haven't captured and compiled the decomposition rules yet,
+        if not self._captured:
+            # Capture the current operations and mark as captured
+            self._captured = True
+
+            # Solve the decomposition graph to get the decomposition rules
+            # for all the captured operations
+            # I know it's a bit hacky to do this here, but it's the only
+            # place where we can be sure that we have seen all operations
+            # in the circuit before the measurement.
+            # TODO: Find a better way to do this.
+            self._decomp_graph_solution = _solve_decomposition_graph(
+                self._operations,
+                self._gate_set,
+                fixed_decomps=self._fixed_decomps,
+                alt_decomps=self._alt_decomps,
+            )
+
+            # Create decomposition rules for each operation in the solution
+            # and compile them to Catalyst JAXPR decomposition rules
+            for op, rule in self._decomp_graph_solution.items():
+                # Get number of wires if exists
+                op_num_wires = (
+                    op.op.params.get("num_wires", None) if hasattr(op.op, "params") else None
+                )
+
+                if (
+                    o := next(
+                        (
+                            o
+                            for o in self._operations
+                            if o.name == op.op.name and len(o.wires) == op_num_wires
+                        ),
+                        None,
+                    )
+                ) is not None:
+                    _create_decomposition_rule(
+                        rule,
+                        op_name=op.op.name,
+                        num_wires=len(o.wires),
+                        requires_copy=self.compiler_ops_num_wires[op.op.name] == -1,
+                    )
+                elif op.op.name in self.compiler_ops_num_wires:
```

Comment:
Can we pre-compile all the decomposition rules for operators with a fixed number of wires and save them somewhere? They will always be the same so they probably don't need to be recompiled for every execution? So when we see one of these operators, instead of compiling the selected decomposition rule, we pick the selected rule from the library of pre-compiled decomposition rules? Just a thought.